File: src/main/scala/wdfeer/avarus/CommandHelper.scala
package wdfeer.avarus

import com.mojang.brigadier.{Command, CommandDispatcher}
import com.mojang.brigadier.builder.LiteralArgumentBuilder.literal
import com.mojang.brigadier.context.CommandContext
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text
import wdfeer.avarus

import scala.language.postfixOps
import wdfeer.avarus.CommandResult.Failure

def registerMessageCommand(
    name: String,
    makeMessage: CommandContext[ServerCommandSource] => Text
)(implicit dispatcher: CommandDispatcher[ServerCommandSource]): Unit = {
  dispatcher.register(literal(name).executes(context => {
    context.getSource.sendMessage(makeMessage(context))
    0
  }))
}

def toCommand(execute: ServerPlayerEntity => CommandResult): Command[ServerCommandSource] = {
  context => {
    val player = context.getSource.getPlayer
    if player != null then {
      val result = execute(player)
      result match
        case CommandResult.Failure(error) => context.getSource.sendMessage(Text.of(error))
        case CommandResult.Success => ()
      
      result.number
    } else 1
  }
}

enum CommandResult(val number: Int) {
  case Success extends CommandResult(0)
  case Failure(error: String) extends CommandResult(1)
}

File: src/main/kotlin/wdfeer/avarus/CommandHelper.kt
package wdfeer.avarus

import com.mojang.brigadier.Command
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext
import net.minecraft.server.command.CommandManager.literal
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text

fun getResultNumber(success: Boolean) = if (success) Command.SINGLE_SUCCESS else 0

sealed class CommandResult(val success: Boolean) {
    val number: Int get() = getResultNumber(success)

    object SilentSuccess : CommandResult(true)
    data class Success(val info: String) : CommandResult(true)
    data class Failure(val error: String) : CommandResult(false)
}

/** Registers a command that sends a message to the player, always succeeds. */
fun registerMessageCommand(
    dispatcher: CommandDispatcher<ServerCommandSource>,
    name: String,
    makeMessage: (CommandContext<ServerCommandSource>) -> Text
) {
    dispatcher.register(literal(name).executes { context ->
        context.source.sendMessage(makeMessage(context))
        0
    })
}

/** Registers a command that performs an action on the player and sends an error message on error. */
fun toCommand(execute: (ServerPlayerEntity) -> CommandResult): Command<ServerCommandSource> {
    return Command { context ->
        val player = context.source.player
        if (player != null) {
            val result = execute(player)
            when (result) {
                CommandResult.SilentSuccess -> {}
                is CommandResult.Success -> context.source.sendMessage(Text.of(result.info))
                is CommandResult.Failure -> context.source.sendMessage(Text.of(result.error))
            }
            result.number
        } else {
            getResultNumber(false)
        }
    }
}